content/05_Filtering.md

```diff
@@ -1,269 +1,314 @@
-    <dd>
-        Reduce the source sequence to a sequence with at most the same number of elements</dd>
-    <dt>Aggregation operators</dt>
-    <dd>
-        Reduce the source sequence to a sequence with a single element</dd>
-    <dt>Fold operators</dt>
-    <dd>
-        Reduce the source sequence to a single element as a scalar value</dd>
-</dl>
+```csharp
+public static class SampleExtensions
+{
+    public static void Dump<T>(this IObservable<T> source, string name)
+    {
+        source.Subscribe(
+            value =>Console.WriteLine($"{name}-->{value}"), 
+            ex => Console.WriteLine($"{name} failed-->{ex.Message}"),
+            () => Console.WriteLine($"{name} completed"));
+    }
+}
+```
 
-We discovered that the creation of an observable sequence from a scalar value is defined as _anamorphism_ or described as an _'unfold'_. We can think of the anamorphism from `T` to `IObservable<T>` as an 'unfold'. This could also be referred to as "entering the monad" where in this case (and for most cases in this book) the monad is `IObservable<T>`. What we will now start looking at are methods that eventually get us to the inverse which is defined as _catamorphism_ or a `fold`. Other popular names for fold are 'reduce', 'accumulate' and 'inject'.
+## Where
 
-## Where								
-Applying a filter to a sequence is an extremely common exercise and the most common filter is the `Where` clause. In Rx you can apply a where clause with the `Where` extension method. For those that are unfamiliar, the signature of the `Where` method is as follows:
+Applying a filter to a sequence is an extremely common exercise and the most straightforward filter in LINQ is the `Where` operator. As usual with LINQ, Rx provides its operators in the form of extension methods. If you are already familiar with LINQ, the signature of Rx's `Where` method will come as no surprise:
 
-    IObservable<T> Where(this IObservable<T> source, Fun<T, bool> predicate)
+```cs
+IObservable<T> Where<T>(this IObservable<T> source, Func<T, bool> predicate)
+```
+
+Note that the element type is the same for the `source` parameter as it is for the return type. This is because `Where` doesn't modify elements—it can filter some out, but those that it does not remove are passed through unaltered.
 
-Note that both the source parameter and the return type are the same. This allows for a fluent interface, which is used heavily throughout Rx and other LINQ code. In this example we will use the `Where` to filter out all even values produced from a `Range` sequence.
+This example uses `Where` to filter out all odd values produced from a `Range` sequence.
 
 ```csharp
-var oddNumbers = Observable.Range(0, 10)
-    .Where(i => i % 2 == 0)
-    .Subscribe(
-        Console.WriteLine, 
-        () => Console.WriteLine("Completed"));
+IObservable<int> xs = Observable.Range(0, 10); // The numbers 0-9
+
+IObservable<int> oddNumbers = xs.Where(i => i % 2 == 0);
```

Comment:
Are you absolutely convinced this produces odd numbers?

Reply:
Utterly, but not necessarily correctly.

Reply:
Actually... having re-read the text, I now believe what I wrote is correct. I said this will:

> filter out all odd values

And that's what it does. It filters those out, leaving only the even ones.

So as far as I can tell, I never said it would _produce_ odd numbers. But I'll add a clarifying "...leaving only the even ones".

Reply:
Oh wait, it's the variable name... 🤦